crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -288,41 +209,706 @@ impl CoinReadApiServer for CoinReadApi {
+                previous_transaction: coin.previous_transaction,
+            })
+            .collect::<Vec<_>>())
+    }
+
+    async fn get_executed_transaction_and_effects(
+        &self,
+        digest: TransactionDigest,
+    ) -> SuiResult<(VerifiedTransaction, TransactionEffects)> {
+        self.state
+            .get_executed_transaction_and_effects(digest)
+            .await
+    }
+
+    async fn get_balance(&self, owner: SuiAddress, coin_type: TypeTag) -> SuiResult<TotalBalance> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_balance(owner, coin_type)
+            .await
+    }
+
+    async fn get_all_balance(
+        &self,
+        owner: SuiAddress,
+    ) -> SuiResult<Arc<HashMap<TypeTag, TotalBalance>>> {
+        self.state
+            .indexes
+            .as_ref()
+            .ok_or(SuiError::IndexStoreNotAvailable)?
+            .get_all_balance(owner)
+            .await
+    }
+}
+
+#[cached(
+    type = "SizedCache<String, ObjectID>",
+    create = "{ SizedCache::with_size(10000) }",
+    convert = r#"{ format!("{}{}", package_id, object_struct_tag) }"#,
+    result = true
+)]
+async fn find_package_object_id(
+    state: Arc<AuthorityState>,
+    package_id: ObjectID,
+    object_struct_tag: StructTag,
+) -> SuiApiResult<ObjectID> {
+    spawn_monitored_task!(async move {
+        let publish_txn_digest = state.find_publish_txn_digest(package_id)?;
+
+        let (_, effect) = state
+            .get_executed_transaction_and_effects(publish_txn_digest)
+            .await?;
+
+        for ((id, _, _), _) in effect.created() {
+            if let Ok(object_read) = state.get_object_read(id) {
+                if let Ok(object) = object_read.into_object() {
+                    if matches!(object.type_(), Some(type_) if type_.is(&object_struct_tag)) {
+                        return Ok(*id);
+                    }
+                }
+            }
+        }
+        Err(SuiRpcInputError::GenericNotFound(format!(
+            "Cannot find object [{}] from [{}] package event.",
+            object_struct_tag, package_id,
+        ))
+        .into())
+    })
+    .await?
+}
+
+#[cfg_attr(test, automock)]
+#[async_trait]
+pub trait CoinReadInternalTrait {
+    fn get_state(&self) -> Arc<dyn StateServiceWrapper + Send + Sync>;
+    async fn get_coins_iterator(
+        &self,
+        owner: SuiAddress,
+        cursor: (String, ObjectID),
+        limit: Option<usize>,
+        one_coin_type_only: bool,
+    ) -> SuiApiResult<CoinPage>;
+    async fn get_coin_metadata(
+        &self,
+        coin_struct: StructTag,
+    ) -> SuiApiResult<Option<SuiCoinMetadata>>;
+    async fn get_treasury_cap(&self, coin_struct: StructTag) -> SuiApiResult<TreasuryCap>;
+}
+
+pub struct CoinReadInternalService {
+    state: Arc<dyn StateServiceWrapper + Send + Sync>,
+    pub metrics: Arc<JsonRpcMetrics>,
+}
```

Comment:
Similar comments to the above:  On naming, what does `Service` mean? Do we need the `dyn` trait here?

Reply:
Ideally, traits and objects that implement traits could share the same name, but since they can't, I settled with `CoinReadInternalService` that implements trait `CoinReadInternal`. This is passed to `CoinReadApi` and captures any logic that doesn't directly involve the rpc layer. The `Arc<dyn` trait is to allow dynamic dispatching, namely so that in our unit tests, we can provide a `MockCoinReadInternalService`. I think the only disadvantage of using dynamic dispatch is potentially more overhead in runtime, but the code is less verbose and we don't need to modify how `CoinReadApi` is instantiated with `new` today

Reply:
I think that's a fair assessment on using trait objects here (I don't think we should be overly concerned about the overhead in this case, but I just wanted to make sure this was a conscious decision.

Re: naming, I'm not a particular fan of it, but adding `Impl` as the suffix is more conventional than `Service` for this, e.g. `CoinReadInternal` and `CoinReadInternalImpl`.  Granted, I don't see that pattern show up anywhere else in our codebase, but I do see `Service` pretty much exclusively meaning something that you talk to over the network (thus my initial confusion).

Reply:
Gotcha, makes sense here and thanks for the suggestion. I'll adopt the `Impl` suffixing